### ISSUE TITLE
Cancel to monitor  topic  repository is not delete topic, next  monitor this topic ,repository has more this  topic

### ISSUE THREAD

**Issue by wuxian12** [php-mqtt/client#47]
Cancel to monitor  topic,  repository is not delete topic, next  monitor this topic ,repository has more this  topic

**Comment by Namoshek** [comment]
I'm slightly confused by your description of the issue, but I guess I know where the problem is.

**TL;DR:** you are not calling `$client->loop()` after `$client->unsubscribe()`.

**Long version:**
PHP works synchronously, meaning we do not have real async functions and stuff other languages like C# bring to the table. For that reason, all MQTT commands which require a response require active _polling of a response_ which is implemented by the `$client->loop()` method. To learn which messages require a response, have a look at the [MQTT v3.1 specification](http://public.dhe.ibm.com/software/dw/webservices/ws-mqtt/mqtt-v3r1.html), especially messages with QoS 1 or higher.

When you use `$client->unsubscribe($topic)`, you are asking the broker to remove you from the recipient list of the given `$topic`. The request is sent immediately. Until the broker responds with an acknowledgement, the client will keep the subscription in the repository though (since it may receive more messages on the topic). Only when the client receives an acknowledgement for the unsubscribe request, which is polled by `$client->loop()`, the client will remove the subscription from the repository.

**Comment by wuxian12** [comment]
thank you，Every time a user logs in, I start a monitor without deleting the monitor. In this case, we have to control whether we have already listened or not

**Comment by Namoshek** [comment]
Yes, that is something you have to handle on your own. The protocol itself does not care about multiple subscriptions to the same or a similar topic from a single client, while some brokers may actually only send a message to the client once, even if it matches multiple subscriptions.

I'm a bit confused by your description though. Are you using a single MQTT client which subscribes for all logged in users or are you using one MQTT client per user? If the latter is the case, you should not have an issue with the `MemoryRepository` actually.

**Comment by wuxian12** [comment]
I adopted a client, but all users of the website shared this client. A user logged in and started a monitoring with the theme of user account，Can I expose a method that I can get to repository and I'll delete the redundant theme myself

**Comment by Namoshek** [comment]
If you share a single client for all of your users, can't you just hold a local list of subscribed themes and only add a new subscription if the theme is actually new?

```php
class MqttClientWrapper
{
    private array $subscribedThemes = [];

    public function subscribe(string $theme, callable $closure): void
    {
        if (in_array($theme, $this->subscribedThemes)) {
            return;
        }

        $this->subscribedThemes[] = $theme;

        // subscribe ...
    }
}
```

**Comment by wuxian12** [comment]
好的谢谢
